### PR TITLE
.Net Samples Restructuring - Phase 2

### PR DESCRIPTION
**Pull request by RogerBarreto**
## Phase 2 - Sample Restructuring

- Removal of Examples numbering 
- Decomposing `KernelSyntaxExamples` into meaningful Feature base folders within `Concepts`.
- Update current sample projects to use `sample` centric `InternalUtilities` reducing code repetition
- Update SDK projects type to be correctly idenfified as xUnit Test projects
- Update projects to be implicit using friendly, reducing considerably the extra lines of code for every sample.
- Decomposing `Concepts\AgentSyntax` into `Concepts\Agents`
- Decomposing `Concepts\AgentSyntax\GettingStarted` into a new root dedicated `GettingStartedWithAgents`

### CODE REVIEW

**Review by crickman — APPROVED** [review]

File: dotnet/samples/Concepts/Memory/HuggingFace_EmbeddingGeneration.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+        var embeddingGenerator = kernel.GetRequiredService<ITextEmbeddingGenerationService>();
+
+        // Generate embeddings for each chunk.
+        var embeddings = await embeddingGenerator.GenerateEmbeddingsAsync(["John: Hello, how are you?\nRoger: Hey, I'm Roger!"]);
+
+        this.WriteLine($"Generated {embeddings.Count} embeddings for the provided text");
+    }
+
+    public HuggingFace_EmbeddingGeneration(ITestOutputHelper output) : base(output)
```
Comment (markwallace-microsoft) [inline dotnet/samples/Concepts/Memory/HuggingFace_EmbeddingGeneration.cs:34]: Can we make this a primary constructor

File: dotnet/samples/Concepts/LocalModels/HuggingFace_ChatCompletionWithTGI.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            if (role is null)
+            {
+                role = chatMessageChunk.Role;
+                Write(role);
+            }
+            Write(chatMessageChunk.Content);
+        }
+    }
+
+    public HuggingFace_ChatCompletionWithTGI(ITestOutputHelper output) : base(output)
```
Comment (markwallace-microsoft) [inline dotnet/samples/Concepts/LocalModels/HuggingFace_ChatCompletionWithTGI.cs:90]: Can we make this a primary constructor

File: dotnet/samples/Concepts/Properties/AssemblyInfo.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+[assembly: System.CLSCompliantAttribute(false)]
+[assembly: Microsoft.Extensions.Configuration.UserSecrets.UserSecretsIdAttribute("5ee045b0-aea3-4f08-8d31-32d1a6f8fed0")]
+[assembly: AssemblyTitle("Concepts")]
+[assembly: AssemblyDescription("Semantic Kernel Concepts Samples Project")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("Microsoft")]
+[assembly: AssemblyProduct("Concepts")]
+[assembly: AssemblyCopyright("Copyright ©  2024")]
+[assembly: AssemblyVersion("1.0.0.0")]
+[assembly: AssemblyFileVersion("1.0.0.0")]
```
Comment (markwallace-microsoft) [inline dotnet/samples/Concepts/Properties/AssemblyInfo.cs:18]: Elsewhere we use the csproj should this be different?

Should we have a hardcoded version here? I'd expect this to get versioned the same way all of the other assemblies are.

File: dotnet/samples/Concepts/TextGeneration/HuggingFace_TextGeneration.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+        Kernel kernel = Kernel.CreateBuilder()
+            .AddHuggingFaceTextGeneration(
+                model: TestConfiguration.HuggingFace.ModelId,
+                apiKey: TestConfiguration.HuggingFace.ApiKey)
+            .Build();
+
+        var questionAnswerFunction = kernel.CreateFunctionFromPrompt("Question: {{$input}}; Answer:");
+
+        var result = await kernel.InvokeAsync(questionAnswerFunction, new() { ["input"] = "What is New York?" });
```
Comment (markwallace-microsoft) [inline dotnet/samples/Concepts/TextGeneration/HuggingFace_TextGeneration.cs:32]: In my recent tests I always had to set the `max_new_tokens` to get a valid response. Why is this not needed here?

File: dotnet/samples/Concepts/TextGeneration/HuggingFace_TextGeneration.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+        WriteLine(result.GetValue<string>());
+    }
+
+    [RetryFact(typeof(HttpOperationException))]
+    public async Task RunStreamingExampleAsync()
+    {
+        WriteLine("\n======== HuggingFace zephyr-7b-beta streaming example ========\n");
+
+        const string Model = "HuggingFaceH4/zephyr-7b-beta";
```
Comment (markwallace-microsoft) [inline dotnet/samples/Concepts/TextGeneration/HuggingFace_TextGeneration.cs:42]: Should we read this from config and provide this as a default suggestion there?

**Review by markwallace-microsoft — COMMENTED** [review]
Added a few minor comments that should be quick to fix

File: dotnet/samples/Concepts/TextGeneration/HuggingFace_TextGeneration.cs (hunk 5 above)
Comment (RogerBarreto) [reply dotnet/samples/Concepts/TextGeneration/HuggingFace_TextGeneration.cs:42]: Not all hugging face public inference API supports Streaming, I added this as a fixed for ensuring a working example, I can add a config and use this as a default. This is the original code, I can change it.

File: dotnet/samples/Concepts/TextGeneration/HuggingFace_TextGeneration.cs (hunk 4 above)
Comment (RogerBarreto) [reply dotnet/samples/Concepts/TextGeneration/HuggingFace_TextGeneration.cs:32]: It depends on the model configured.

Using: `HuggingFaceH4/zephyr-7b-beta` I get this output.
image

Added the remarks with details about the tested as well as defaulting to one of the successful tested models.

File: dotnet/samples/Concepts/Properties/AssemblyInfo.cs (hunk 3 above)
Comment (RogerBarreto) [reply dotnet/samples/Concepts/Properties/AssemblyInfo.cs]: I reverted this change, I was having a conflict in my local build got it sorted, and removed this file completely.

**Review by markwallace-microsoft — APPROVED** [review]

**Pull request merged**